File: A1779617_KaifengChen/functions/src/MachineLearning/Data2TrainTest.ts
/* eslint-disable require-jsdoc */
/* eslint-disable max-len */
import * as fs from "fs";
import * as path from "path";

export function Data2TrainTest(sourceDir: string, trainTargetDir: string, testTargetDir: string, trainRatio: number): void {
  const FOLDER_NAME = "data";
  trainTargetDir = path.join(FOLDER_NAME, trainTargetDir);
  testTargetDir = path.join(FOLDER_NAME, testTargetDir);

  if (!fs.existsSync(trainTargetDir)) {
    fs.mkdirSync(trainTargetDir, {recursive: true});
  }

  if (!fs.existsSync(testTargetDir)) {
    fs.mkdirSync(testTargetDir, {recursive: true});
  }

  const persons = fs.readdirSync(sourceDir).filter((dir) => fs.lstatSync(path.join(sourceDir, dir)).isDirectory());

  for (const person of persons) {
    const images = fs.readdirSync(path.join(sourceDir, person)).filter((file) => [".png", ".jpg", ".jpeg"].includes(path.extname(file)));
    const splitIdx = Math.floor(trainRatio * images.length);
    const trainImages = images.slice(0, splitIdx);
    const testImages = images.slice(splitIdx);

    const trainPersonDir = path.join(trainTargetDir, person);
    const testPersonDir = path.join(testTargetDir, person);

    if (!fs.existsSync(trainPersonDir)) {
      fs.mkdirSync(trainPersonDir);
    }

    if (!fs.existsSync(testPersonDir)) {
      fs.mkdirSync(testPersonDir);
    }

    trainImages.forEach((img, idx) => {
      const newImgName = `${idx}${path.extname(img)}`;
      fs.renameSync(path.join(sourceDir, person, img), path.join(trainPersonDir, newImgName));
    });

    testImages.forEach((img, idx) => {
      const newImgName = `${idx}${path.extname(img)}`;
      fs.renameSync(path.join(sourceDir, person, img), path.join(testPersonDir, newImgName));
    });
  }

  console.log("[INFO] Data split into training and testing sets.");
}

File: A1869202_Revanth/Week10/Data2TrainTest.py
# This converts the given folder images data to train and test data.
import os
import shutil
from random import shuffle
from pathlib import Path

def Data2TrainTest(source_dir:Path, train_target_dir:Path, test_target_dir:Path, train_ratio:float):
    # Create train and test directories if they don't exist
    # save them under data/train_target_dir, data/test_target_dir folders
    FolderName = "data/"
    train_target_dir = os.path.join(FolderName, train_target_dir)
    test_target_dir = os.path.join(FolderName, test_target_dir)
    os.makedirs(train_target_dir, exist_ok=True)
    os.makedirs(test_target_dir, exist_ok=True)

    # List all the person folders
    persons = [d for d in os.listdir(source_dir) if os.path.isdir(os.path.join(source_dir, d))]

    # only keep randomly selected 10 classes
    shuffle(persons)

    print("[INFO] Splitting data into Train & Test...")
    for person in (persons):
        # List all images of the person
        images = [f for f in os.listdir(os.path.join(source_dir, person)) if f.endswith(('.png', '.jpg', '.jpeg'))]
        shuffle(images)

        # Calculate split index
        split_idx = int(train_ratio * len(images))
        train_images = images[:split_idx]
        test_images = images[split_idx:]

        # Define person-specific directories in train and test
        train_person_dir = os.path.join(train_target_dir, person)
        test_person_dir = os.path.join(test_target_dir, person)

        os.makedirs(train_person_dir, exist_ok=True)
        os.makedirs(test_person_dir, exist_ok=True)

        # Rename images to respective directories, a copy of faces in the train and test folders
        for idx, img in enumerate(train_images):
            new_img_name = str(idx) + os.path.splitext(img)[-1]  # new name with extension
            # Create a copy of the image in the train folder
            shutil.copy(os.path.join(source_dir, person, img), os.path.join(train_person_dir, new_img_name))

        for idx, img in enumerate(test_images):
            new_img_name = str(idx) + os.path.splitext(img)[-1]  # new name with extension
            shutil.copy(os.path.join(source_dir, person, img), os.path.join(test_person_dir, new_img_name))
    
    print(f"[INFO] Splitting Done: {len(persons)} persons data splitted into {train_ratio*100:.1f}% train and {(1-train_ratio)*100:.1f}% test data")

    return FolderName